src/Sylius/Behat/Context/Api/Shop/ProductContext.php

```diff
@@ -661,4 +675,26 @@ private function hasProductWithMainImage(): bool
 
         return $images[0]['type'] === 'main' && $images[0]['path'];
     }
+
+    private function productHasAssociationsWithProducts(array $associations, string $productAssociationName, array $products): bool
+    {
+        foreach ($associations as $associationIri) {
+            if ($this->responseChecker->hasValue($this->client->showByIri($associationIri), 'type', sprintf('/api/v2/admin/product-association-types/%s', StringInflector::nameToLowercaseCode($productAssociationName)))) {
+               foreach ($products as $product) {
+                   if ($this->productIsAssociated($product, $associationIri)) {
+                       return true;
+                   }
+               }
+            }
+        }
+
+        return false;
+    }
+
+    private function productIsAssociated(ProductInterface $product, string $associationIri): bool
+    {
+        $response = $this->iriConverter->getIriFromItem($product);
+
+        return in_array($response, $this->responseChecker->getValue($this->client->showByIri($associationIri), 'associatedProducts'));
```

Comment:
This is the second time you are doing `$this->client->showByIri($associationIri)` to check the same thing (the first was in `productHasAssociationsWithProducts`), refactor these private methods to call this endpoint once